### DISCUSSION TITLE
Performance of Uvicorn+Gunicorn

### DISCUSSION THREAD

**Discussion by otbutz** [osm-search/Nominatim#3097] [General]
I stumbled upon your blog post benchmarking the new Python codebase: https://nominatim.org/2023/04/03/reverse-performance.html

> Python is tested with three different web frameworks: [starlette](https://www.starlette.io/), [sanic](https://sanic.dev/) and [falcon](https://falcon.readthedocs.io/) starlette and falcon are served with [uvicorn](http://www.uvicorn.org/) fronted by nginx.

Did you also test it in combination with gunicorn?

https://www.uvicorn.org/deployment/#gunicorn

**Comment by lonvia** [comment]
Yes, I tried gunicorn as well. It made no difference at all.

I also tried different setups how nginx and uvicorn using one global socket or one socket per workers. I tested unix sockets vs. http sockets. nginx is better able to distribute load over multiple workers. But again none of this made any difference to the final performance. The bottleneck is clearly with SQLAlchemy and the database.

I just recently became aware of this tweet by the developer of psycopg from December: https://twitter.com/psycopg/status/1606390672277311501 where he complains about threads inexplicably slowing down each other. The symptoms look very similar. So it might be something to follow up on.

**Comment by otbutz** [comment]
Thanks for the detailed response. Good to know that the bottleneck is already identified.

https://github.com/psycopg/psycopg/issues/448

**Reply by lonvia** [reply]
That's not exactly the root cause then. :( There is no difference between using asyncpg and psycopg for Nominatim. In fact, the latter fares slightly better.

**Reply by otbutz** [reply]
Hmm. If it's all CPU bound, I'd throw a profiler at it to find the culprit.
